pkg/client/mapper.go

```diff
@@ -103,3 +103,57 @@ func NewBidirectionalNewOneToManyGroupMapper(source, dest, groupMappingFile stri
 	}
 	return nil, nil, fmt.Errorf("unsupported source to dest mapper type: source %s, dest %s", source, dest)
 }
+
+// UserMapperImpl implements groupsync.UserMapper.
+type UserMapperImpl map[string]string
+
+func (u UserMapperImpl) MappedUserID(ctx context.Context, userID string) (string, error) {
+	v, ok := u[userID]
+	if !ok {
+		return "", groupsync.ErrTargetUserIDNotFound
+	}
+	return v, nil
+}
+
+// NewGoogleGroupGitHubUserMapper creates a UserMapperImpl that maps
+// google user email to github user handle.
+func NewGoogleGroupGitHubUserMapper(userMappingFile string) (groupsync.UserMapper, error) {
+	b, err := os.ReadFile(userMappingFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read mapping file: %w", err)
+	}
+	tm := &v1alpha3.UserMappings{}
+	if err := prototext.Unmarshal(b, tm); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal mapping file: %w", err)
+	}
+
+	ggToGHUserMapping := make(UserMapperImpl)
+	ghToGGUserMapping := make(UserMapperImpl)
+
+	for _, mapping := range tm.GetMappings() {
+		src, dst := mapping.GetGoogleUserEmail(), mapping.GetGitHubUserId()
+		// skip user if he doesn't have google group or github that needs mappings.
+		if src == "" || dst == "" {
+			continue
+		}
+		// Check user mapping relation is 1:1.
+		if existingDst, ok := ggToGHUserMapping[src]; ok && existingDst != dst {
+			return nil, fmt.Errorf("google group user %s mapped to multiple github user %s,%s", src, existingDst, dst)
+		}
```

Comment:
I was thinking use this as a validation to make sure the mapping is 1:1, so it can catching mistakes in the mapping file before we start the membership syncing.

Reply:
But why is that a problem? I'm trying to understand from a _product_ perspective why this is something worthy of stopping all processing on.

Reply:
My thought on this is the old map value will be overwritten by the new one, and causing only one of them can be synced, user might not realize that if they mistakenly put two entries with same src but different dst. I think this is worthing letting user know.

Reply:
It's still not clear what action you expect the user to take? We didn't tell them where the duplication exists or how to resolve it. This error feels like a classic "an error occurred" message.

Reply:
What do you think if I update the error message to be 
`google group user abc mapped to multiple github user foo, bar`, please make sure the it's a 1:1 mapping relation in your textproto tile.`

Or do you think removing this check would make more sense.

Reply:
I think that message is correct. I'm fine leaving it as an error for now, but we may want to consider downgrading it to a warning in the future.

Reply:
Updated to `logger.WarnContext` on duplicated info.